Add Ethernet support for NXP S32Z/E SoCs

manuargue opened the pull request:
This PR introduces support for Ethernet on NXP S32Z/E SoCs, specifically for S32Z27x-DC(2) board.

The Ethernet low-level driver is based on NXP S32 Network Controller (NETC) complex, which is composed of an External MDIO controller, partitionable Station Interfaces (SI) and an integrated Ethernet Switch. Current driver allows to manage from Zephyr a Physical Station Interface (PSI) and/or a Virtual SI (VSI). The rest of the VSI's shall be assigned to different cores of the system. The Switch is initialized from the Ethernet shim driver with a default configuration, and all ports are enabled and are transparent for the user. A separate Switch driver will be addressed in a future PR as needed.

NETC uses MSI-X, therefore a new Mbox-based driver is implemented for the Message Receive Unit (MRU) peripheral, to be able to terminate in interrupts targetting the core controlling the SI.

manuargue-nxp committed:
drivers: mbox: separate NRFx specific Kconfig

Separate NRFx Mbox driver Kconfig to make place for other drivers.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: mbox: add support for NXP S32 MRU

Introduce NXP S32 Message Receive Unit (MRU) driver based
of Mbox API. The MRU couples with a processor and allows to
receive messages from senders, which are other modules or
processors.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
dts: arm: nxp_s32z27x_r52: add MRU instances

Add MRU instances to be used with NXP S32 MRU Mbox based driver.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: eth: phy_mii: support Gigabit ethernet

Advertise Gigabit Ethernet if the PHY supports it. As with the
other speeds, it is assumed the PHY supports both duplex modes.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: mdio: add support for NXP S32 NETC EMDIO

Introduce NXP S32 NETC External MDIO controller driver. Driver supports
a single instance, as current support is based on NXP S32Z/E SoCs.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: s32z270dc2_r52: enable Ethernet support

Introduce DT nodes for NETC complex and enable its usage for
s32z270dc2_r52 boards. Using PSI0 as default networking interface and
Switch Port0 as it's the only port available on this board.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: ethernet: add support for NXP S32 NETC

Introduce Ethernet low-level driver for NXP S32 Network Controller
(NETC). Current driver allows to manage from Zephyr a Physical Station
Interface (SI) and/or a Virtual SI. The NETC has an integrated Ethernet
Switch. Currently the Switch is initialized from this driver with a
default configuration, and all ports are enabled and transparent for
the user. A separate Switch driver should be addressed in future patches.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
samples: boards: nxp_s32: add sample for NETC

The sample application shows how to configure NXP S32 Network Controller
(NETC) for different use-cases.

Signed-off-by: Manuel Arguelles <manuel.arguelles@nxp.com>

manuargue commented:
Hello @tbursztyka, @rlubos, @carlocaione, could you please provide some feedback on these changes? Thanks

dleach02 commented:
@manuargue the HAL PR was merged a couple weeks back so I would update the west.yml (or maybe rebase if the HAL changes are already part of main)

manuargue commented:
> @manuargue the HAL PR was merged a couple weeks back so I would update the west.yml (or maybe rebase if the HAL changes are already part of main)

already did that in last rebase/push, but for some reason it didn't remove the DNM label on this pr... re-pushed

manuargue commented:
> @manuargue the HAL PR was merged a couple weeks back so I would update the west.yml (or maybe rebase if the HAL changes are already part of main)

already rebased and removed changes in manifest on previous update since hal_nxp is already up to date, but for some reason the DNM label is not removed. Rebased again.

danieldegrasse commented on drivers/mbox/Kconfig:
```diff
@@ -9,20 +9,16 @@ menuconfig MBOX
 
+# Include these first so that any properties (e.g. defaults) below can be
+# overridden (by defining symbols in multiple locations)
+source "drivers/mbox/Kconfig.nrfx"
```
Minor comment (non-blocking): Is there an MBOX driver that needs to override `CONFIG_MBOX_INIT_PRIORITY`? I assume that this pattern was followed due to the way that the "root" MDIO drivers KConfig file appears to be written, but in my experience the individual driver KConfig files are typically sourced after the primary KConfig file has defined all symbols

laurenmurphyx64 commented:
Ping @tbursztyka. Sorry if this is annoying. I can't merge without the assignee's approval and I thought it'd be helpful to get this back on your docket before feature freeze this Friday.
